### PR TITLE
fix(cross-seed): use correct save path for tracker-named categories in reflink/hardlink mode

### LINKED ISSUE #1703: Bug: Cross-seed creates tracker-named categories with incorrect save paths (reflink/hardlink + by-tracker preset)
## Description

When cross-seed runs with `UseCategoryFromIndexer=true` and reflink/hardlink mode using the `by-tracker` directory preset, newly created categories are assigned incorrect `savePath` values. The category save paths end up pointing to torrent-specific reflink tree directories (e.g., `/data/cross-seed/LST/Movie.Name--0bbfd477`) instead of the tracker-level base directory (e.g., `/data/cross-seed/LST`).

Additionally, when multiple cross-seeds are processed concurrently, a race condition causes category save paths to be shuffled across tracker categories — e.g., the "FearNoPeer" category ends up with LST's path, "LST" gets seedpool's path, etc.

## Steps to reproduce

1. Configure an instance with:
   - Reflinks (or hardlinks) enabled
   - `HardlinkDirPreset` = `by-tracker`
   - `HardlinkBaseDir` includes a cross-seed directory (e.g., `/data/cross-seed/`)
2. Configure cross-seed automation with `UseCategoryFromIndexer = true`
3. Run a dirscan or completion search that matches a torrent across multiple trackers
4. Observe that categories are created with incorrect save paths

## Observed behavior

Categories created by cross-seed have torrent-specific paths:

```json
{
  "FearNoPeer": {
    "name": "FearNoPeer",
    "savePath": "/data/cross-seed/LST/Past.Lives.2023.UHD.BluRay.2160p.DD.5.1.DV.HDR.x265-BHDStudio.mp4--0bbfd477"
  },
  "LST": {
    "name": "LST",
    "savePath": "/data/cross-seed/seedpool (API)/Past.Lives.2023.UHD.BluRay.2160p.DD.5.1.DV.HDR.x265-BHDStudio.mp4--f28fb640"
  },
  "OnlyEncodes+ (API)": {
    "name": "OnlyEncodes+ (API)",
    "savePath": "/data/cross-seed/TorrentLeech/Past.Lives.2023.UHD.BluRay.2160p.DD.5.1.DV.HDR.x265-BHDStudio.mp4--8ec1f765"
  }
}
```

Note the paths are also shuffled across categories (FearNoPeer has LST's path, etc.).

## Expected behavior

Categories should be created with the tracker-level base directory as their save path:

```json
{
  "FearNoPeer": {
    "name": "FearNoPeer",
    "savePath": "/data/cross-seed/FearNoPeer"
  },
  "LST": {
    "name": "LST",
    "savePath": "/data/cross-seed/LST"
  },
  "OnlyEncodes+ (API)": {
    "name": "OnlyEncodes+ (API)",
    "savePath": "/data/cross-seed/OnlyEncodes+ (API)"
  }
}
```

## Root cause

Two issues in `internal/services/crossseed/service.go`:

### 1. Category save path determined too early with wrong fallback

At lines ~4420-4456, `categorySavePath` is determined *before* entering reflink/hardlink processing. The logic is:
- Look up `baseCategory` in qBittorrent's existing categories
- If not found (it won't be — it's a new tracker-named category), fall back to `props.SavePath` (the matched torrent's save path, e.g., `/data/movies`)

This fallback is incorrect for `by-tracker` mode. The correct save path is `{selectedBaseDir}/{trackerName}`, but `selectedBaseDir` isn't known until inside `processReflinkMode()`/`processHardlinkMode()` where `FindMatchingBaseDir()` and `buildHardlinkDestDir()` are called.

### 2. Concurrent processing causes path shuffling

`ensureCrossCategory()` uses `singleflight` and a `createdCategories` sync.Map cache (lines 10015-10026). When multiple cross-seeds for different trackers process concurrently, the first goroutine to reach `ensureCrossCategory` for a given category creates it with whatever `categorySavePath` it computed. Due to the incorrect fallback in issue #1, concurrent goroutines racing through the category creation can associate the wrong tracker's path with the wrong category.

### 3. Existing categories with wrong paths are never corrected

`ensureCrossCategory()` (line 10037-10048) only logs a warning when an existing category has a different save path — it never updates it. So once a category is created with a wrong path, it stays wrong.

## Proposed fix

**Move category creation into the reflink/hardlink processing functions** where the correct path components are already available.

Inside `processReflinkMode()` and `processHardlinkMode()`, after `selectedBaseDir` and `trackerDisplayName`/`incomingTrackerDomain` are resolved (and before the torrent is added), compute the correct category save path:

```go
// For by-tracker preset: category save path is baseDir/trackerName
// For by-instance preset: category save path is baseDir/instanceName  
// For flat preset: category save path is baseDir
categorySavePath := selectedBaseDir
if instance.HardlinkDirPreset == "by-tracker" {
    trackerDisplayName := s.resolveTrackerDisplayName(ctx, incomingTrackerDomain, req)
    categorySavePath = filepath.Join(selectedBaseDir, pathutil.SanitizePathSegment(trackerDisplayName))
} else if instance.HardlinkDirPreset == "by-instance" {
    categorySavePath = filepath.Join(selectedBaseDir, pathutil.SanitizePathSegment(candidate.InstanceName))
}

if crossCategory != "" {
    if err := s.ensureCrossCategory(ctx, candidate.InstanceID, crossCategory, categorySavePath); err != nil {
        // handle error...
    }
}
```

The early category creation block (lines ~4429-4456) should be skipped when reflink or hardlink mode will be used, since those code paths will handle it with the correct save path.

For `ensureCrossCategory()`, keep the existing warn-only behavior for mismatched save paths on existing categories to avoid breaking users' existing setups. The fix prevents incorrect paths from being set going forward.

## Impact

- Categories created with wrong save paths cause confusion in the qBittorrent UI
- If a user subsequently enables AutoTMM on these categories, torrents would be moved to the wrong location
- The shuffled paths make it difficult to understand which tracker's content is where

## Environment

- Instance with local filesystem access, **reflinks enabled**
- `HardlinkDirPreset`: `by-tracker`
- Cross-seed automation with `UseCategoryFromIndexer`: `true`
- Multiple trackers matching the same content concurrently

### PR DESCRIPTION
**Pull request by nitrobass24**
## Summary

- Defer category creation to reflink/hardlink processing functions where the correct base directory and tracker name are resolved
- Compute category save path as `{baseDir}/{trackerName}` instead of falling back to the matched torrent's save path
- Handle fallback to regular mode when reflink/hardlink handlers return `Used=false` — ensure category is created before proceeding
- Track and report category creation failures in hardlink/reflink result messages

## Test plan
- [x] New unit tests for `buildCategorySavePath` covering all presets (by-tracker, by-instance, flat, unknown, empty), edge cases (trailing slash, special characters, no tracker info), and isolation folder exclusion
- [ ] Manual test with reflink + by-tracker preset + `UseCategoryFromIndexer=true` across multiple trackers
- [ ] Verify categories get correct `{baseDir}/{trackerName}` save paths, not torrent-specific paths
- [ ] Verify concurrent cross-seeds don't shuffle save paths across categories

Fixes #1703

## Summary by CodeRabbit

* **New Features**
  * Improved category save path handling based on hardlink directory preset configuration (by-tracker, by-instance, or flat mode).

* **Bug Fixes**
  * Enhanced category isolation status messaging to indicate when category isolation fails in hardlink/reflink operations.
  * Fixed category creation logic to properly defer and retry category creation in reflink and hardlink modes when needed.

### CODE REVIEW

File: internal/services/crossseed/service.go [hunk 1]
```diff
@@ -4445,14 +4445,20 @@ func (s *Service) processCrossSeedCandidate(
...
+		if !useReflinkMode && !useHardlinkMode {
+			// Ensure the cross-seed category exists with the correct SavePath
+			if err := s.ensureCrossCategory(ctx, candidate.InstanceID, crossCategory, categorySavePath); err != nil {
+				log.Warn().Err(err).
+					Str("category", crossCategory).
+					Str("savePath", categorySavePath).
+					Msg("[CROSSSEED] Failed to ensure category exists, continuing without category")
+				crossCategory = ""            // Clear category to proceed without it
+				categoryCreationFailed = true // Track for result message
+			}
```
[the remark this answers is not in the document]
Comment (nitrobass24) [reply internal/services/crossseed/service.go:4461]: This is already fixed in the current code:

  1. Line 4452: if !useReflinkMode && !useHardlinkMode — category creation is skipped when link modes are enabled
  2. Inside processReflinkMode/processHardlinkMode — category is created with the correct path derived from selectedBaseDir
  3. Lines 4497-4508: If link modes returned Used=false, category is created with the original categorySavePath (the matched torrent's path, which is correct for regular mode)

  The concern about ensureCrossCategory/createdCategories/Used interacting badly doesn't apply — when Used=false, the link-mode function exited before reaching ensureCrossCategory, so nothing was cached in the sync.Map.

**Comment by s0up4200** [comment]
@nitrobass24 you can create branches now, so no need to fork 👍 That way we get docker builds as well!

Are you on our discord btw?

**Comment by nitrobass24** [comment]
Ok yea I wasnt sure of the protcol, there so just figured i'd stay on my fork. But will move over after I fix this issue. No im not on discord, where do I find the link?

**Comment by s0up4200** [comment]
> Ok yea I wasnt sure of the protcol, there so just figured i'd stay on my fork. But will move over after I fix this issue. No im not on discord, where do I find the link?

https://discord.autobrr.com/qui 👍 Ping me in one of the channels and I'll add you in the dev channel. This is only if you want to of course, no pressure if you're not using Discord or anything

**Comment by nitrobass24** [comment]
Reopening from upstream branch
